llrt_core/src/modules/http/url.rs

```diff
@@ -470,29 +326,69 @@ pub fn file_url_to_path<'js>(ctx: Ctx<'js>, url: Value<'js>) -> Result<String> {
 }
 
 pub fn url_format<'js>(url: Class<'js, URL<'js>>, options: Opt<Value<'js>>) -> Result<String> {
-    let mut fragment = true;
-    let mut unicode = false;
-    let mut auth = true;
-    let mut search = true;
+    let mut include_fragment = true;
+    let mut unicode_encode = false;
+    let mut include_auth = true;
+    let mut include_search = true;
 
     // Parse options if provided
     if let Some(options) = options.0 {
-        if options.is_object() {
-            let options = options.as_object().unwrap();
-            if let Some(value) = options.get("fragment")? {
-                fragment = value;
+        if let Some(options) = options.as_object() {
+            if let Ok(value) = options.get("fragment") {
+                include_fragment = value;
             }
             if let Ok(value) = options.get("unicode") {
-                unicode = value;
+                unicode_encode = value;
             }
             if let Ok(value) = options.get("auth") {
-                auth = value;
+                include_auth = value;
             }
             if let Ok(value) = options.get("search") {
-                search = value
+                include_search = value
+            }
+        }
+    }
+
+    let url = url.borrow();
+    let search = if include_search {
+        url.search()
+    } else {
+        String::from("")
+    };
+
+    let hash = if include_fragment {
+        url.hash()
+    } else {
+        String::from("")
+    };
+
+    let mut user_info = String::new();
+    if include_auth {
+        let username = url.username();
+        let password = url.password();
+        if !username.is_empty() {
+            user_info.push_str(&username);
+            if !password.is_empty() {
+                user_info.push(':');
+                user_info.push_str(&password)
             }
+            user_info.push('@')
         }
     }
 
-    Ok(url.borrow().format(auth, fragment, search, unicode))
+    let host = if unicode_encode {
+        domain_to_unicode(&url.host())
+    } else {
+        url.host()
+    };
+
+    Ok(format!(
```

Comment:
I know `concat!` is much faster than `format!`, but you can only use `concat!` with literal arguments, correct? At least I get errors when I try to change it.

Reply:
Yeah that's right. Then I maybe join is a good alternative? I try to avoid format unless I actually need formatting.
```rust
Ok([url.protocol(), "//", user_info, host, url.pathname(), search, hash].join(""))
```

Reply:
Ah, great idea!

Reply:
Replaced all the `format!`s with `join`s.